kratos/utilities/normal_calculation_utils.cpp

```diff
@@ -22,6 +23,264 @@
+        for(auto& r_cond: rModelPart.Elements()) {
+            for(auto& r_node: r_cond.GetGeometry()) {
+                noalias(r_node.FastGetSolutionStepValue(NORMAL)) = zero;
+            }
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+void NormalCalculationUtils::CalculateNormals<Condition>(
+    ModelPart& rModelPart,
+    const bool EnforceGenericAlgorithm
+    )
+{
+    // Getting process info
+    const auto& r_process_info = rModelPart.GetProcessInfo();
+
+    // Getting dimension
+    const SizeType dimension = r_process_info.GetValue(DOMAIN_SIZE);
+
+    // Sum all the nodes normals
+    auto& r_conditions_array = rModelPart.Conditions();
+    const auto it_cond_begin = r_conditions_array.begin();
+
+    // Checking if we can compute with simplex
+    const GeometryData::KratosGeometryType geometry_type = it_cond_begin->GetGeometry().GetGeometryType();
+    const bool use_simplex = EnforceGenericAlgorithm ? false : dimension == 2 ? geometry_type == GeometryData::KratosGeometryType::Kratos_Line2D2 : geometry_type == GeometryData::KratosGeometryType::Kratos_Triangle3D3;
+
+    if (use_simplex) {
+        CalculateOnSimplex(rModelPart, dimension);
+    } else {
+        // Initialize the normals
+        InitializeNormals<Condition>(rModelPart);
+
+        // Calculate the normals in the conditions
+        CalculateNormalsInConditions(rModelPart);
+
+        // Declare auxiliar coordinates
+        Point::CoordinatesArrayType aux_coords;
+
+        #pragma omp parallel for firstprivate(aux_coords)
+        for (int i = 0; i < static_cast<int>(r_conditions_array.size()); ++i) {
+            auto it_cond = it_cond_begin + i;
+            GeometryType& r_geometry = it_cond->GetGeometry();
+
+            // Avoid not "flat" conditions
+            if (r_geometry.WorkingSpaceDimension() != r_geometry.LocalSpaceDimension() + 1) {
+                continue;
+            }
+
+            // Iterate over nodes
+            const double coefficient = 1.0/static_cast<double>(r_geometry.PointsNumber());
+            for (NodeType& r_node : r_geometry) {
+                r_geometry.PointLocalCoordinates(aux_coords, r_node.Coordinates());
+                const array_1d<double, 3> normal = r_geometry.Normal(aux_coords);
+                r_node.SetLock();
+                noalias(r_node.FastGetSolutionStepValue(NORMAL)) += normal * coefficient;
+                r_node.UnSetLock();
+            }
+        }
+
+        // For MPI: correct values on partition boundaries
+        rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+void NormalCalculationUtils::CalculateNormals<Element>(
```

Comment:
why is this different for elements & conditions?

Reply:
Look at the code ad you will see why

Reply:
I did and I don't understand it
All we are interested is Geometries, doen't matter if its an element or a condition

Reply:
how about the following (using names just to show my idea):
Interface:
- `CalculateElementNormals`
- `CalculateConditionNormals`

in the cpp: template<TEntityType> CalculateCondition(Entities)
=> this method is called from `CalculateElementNormals` and `CalculateConditionNormals`

Reply:
This avoids two declarations in the header

Reply:
ping